### PR TITLE
Update to libp2p v0.20.1

### PR DESCRIPTION
**Pull request by twittner**
Several other crates are updated too, e.g. `futures`, `futures_codec`, `yamux`, etc. Please have a look at the `Cargo.lock`.

### CODE REVIEW

**Comment by twittner** [comment]
> Should we test this out on a sentry node beforehand [...]?

Yes.

**Comment by tomaka** [comment]
You forgot the `libp2p-wasm-ext` dependency in the browser tests, which is why CI is failing and the `Cargo.lock` changes are so big.

File: Cargo.lock [hunk 1]
```diff
@@ -9066,7 +9146,7 @@ version = "1.5.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "3bfd5b7557925ce778ff9b9ef90e3ade34c524b5ff10e239c69a42d546d2af56"
 dependencies = [
- "rand 0.7.3",
+ "rand 0.3.23",
```
Comment (tomaka) [inline Cargo.lock:9149]: !

Comment (twittner) [reply Cargo.lock:9149]: Yeah, this happened in the past, [even with single package upgrades](https://github.com/paritytech/substrate/pull/3448).

**Comment by mxinden** [comment]
> You forgot the `libp2p-wasm-ext` dependency in the browser tests, which is why CI is failing

@tomaka  Can you expand on this? `libp2p-wasm-ext` didn't [change any time recently](https://github.com/libp2p/rust-libp2p/blob/master/transports/wasm-ext/Cargo.toml#L3).

Are you suggesting the following patch within Substrate?

```diff
diff --git a/bin/node/browser-testing/Cargo.toml b/bin/node/browser-testing/Cargo.toml
index 25fafbaee..54405faa7 100644
--- a/bin/node/browser-testing/Cargo.toml
+++ b/bin/node/browser-testing/Cargo.toml
@@ -8,7 +8,7 @@ license = "Apache-2.0"
 
 [dependencies]
 futures-timer = "3.0.2"
-libp2p = { version = "0.20.0", default-features = false }
+libp2p = { version = "0.20.0", features = ["libp2p-wasm-ext"], default-features = false }
 jsonrpc-core = "14.2.0"
 serde = "1.0.106"
 serde_json = "1.0.48"
```

**Comment by tomaka** [comment]
I'm talking about this dependency:

https://github.com/paritytech/substrate/blob/e287915eb37928491adbf4666e764ea0be5ac21a/utils/browser/Cargo.toml#L18

The `sc_network::config` module reexports `libp2p_wasm_ext::ExtTransport` specifically for this reason.
It might be a good idea to remove this `libp2p-wasm-ext` dependency and use what `sc_network::config` reexports instead.

**Comment by twittner** [comment]
> I don't understand why the changes in `Cargo.lock` are so big then? Why do we have so many new crates?

As I described [above](https://github.com/paritytech/substrate/pull/6465#issue-437956113) I also updated several other crates which were lagging behind.

**Comment by tomaka** [comment]
I'm thinking about the fact that we *add* 100 new lines to the Cargo.lock that weren't there before.

**Comment by twittner** [comment]
> I'm thinking about the fact that we _add_ 100 new lines to the Cargo.lock that weren't there before.

I think most of these are changes in crypto crates, most likely due to the upgrade of `snow`.

**Comment by twittner** [comment]
I wanted this to be built by CI before going back from WIP to ready for review, but the failing `continuous-integration/gitlab-skip-if-draft` jobs indicates that WIP PRs are not properly built IIUC.

**Comment by tomaka** [comment]
Seems to introduce spikes at the authority-discovery interval again:

Screenshot from 2020-06-24 14-36-04

I don't have the motivation to go through every single CHANGELOG, but the primary suspect would be https://github.com/libp2p/rust-libp2p/pull/1473

**Comment by tomaka** [comment]
We also have a node running https://github.com/libp2p/rust-libp2p/pull/1618, and spikes aren't present on that other node.

**Comment by romanb** [comment]
> Seems to introduce spikes at the authority-discovery interval again [..]

Is there an issue or discussion somewhere on prior occurrences of this problem and how it was resolved (since you write _again_)?

> We also have a node running libp2p/rust-libp2p#1618, and spikes aren't present on that other node.

Since that branch also includes https://github.com/libp2p/rust-libp2p/pull/1473, the problem seems to be elsewhere. I wouldn't expect https://github.com/libp2p/rust-libp2p/pull/1473 to make a big difference in any case, since almost all of the new code (paths) are disabled by default (i.e. the new "feature" of disjoint paths is disabled by default). The only changes that are in `libp2p-0.20.1` but not in https://github.com/libp2p/rust-libp2p/pull/1618 seem to be some dependency updates, which would mean the different behaviour either is related to some such update, or a difference in the substrate code base.

What version / commit of substrate do these two nodes (the one with this branch and the one with https://github.com/libp2p/rust-libp2p/pull/1618) run? Presumably the host machine specifications / hardware for both nodes are the same? Relatedly, on which Grafana Dashboard can one find the graph that you took the screenshot from?

**Comment by tomaka** [comment]
It turns out that these spikes are already present on Polkadot master.

> We also have a node running libp2p/rust-libp2p#1618, and spikes aren't present on that other node.

My apologies. After checking, I had forgotten that I had based my testing upon an older version of Substrate because at that time Polkadot master and Substrate master weren't compatible.

**Comment by tomaka** [comment]
Considering that there isn't any visible difference between before and after this PR, I would be ok for merging this.

**Comment by tomaka** [comment]
Please merge master.

**Comment by sjeohp** [comment]
right now it can't distinguish allowable failures but that will change asap

**Comment by tomaka** [comment]
> right now it can't distinguish allowable failures but that will change asap

Well, the problem here is more the catastrophic CI server failure, not the bot.

**Comment by tomaka** [comment]
bot merge

**Pull request merged**